File: c/ChaCha20/chacha20.c

#include "chacha20.h"
#include <string.h>

#define ROTL(a,b) (((a) << (b)) | ((a) >> (32 - (b))))
#define QR(a, b, c, d) \
    a += b; d ^= a; d = ROTL(d, 16); \
    c += d; b ^= c; b = ROTL(b, 12); \
    a += b; d ^= a; d = ROTL(d, 8); \
    c += d; b ^= c; b = ROTL(b, 7);

static const char constant[16] = "expand 32-byte k";

static uint32_t load32_le(const uint8_t *src) {
    return ((uint32_t)src[0]) |
           ((uint32_t)src[1] << 8) |
           ((uint32_t)src[2] << 16) |
           ((uint32_t)src[3] << 24);
}

static void store32_le(uint8_t *dst, uint32_t val) {
    dst[0] = (uint8_t)(val);
    dst[1] = (uint8_t)(val >> 8);
    dst[2] = (uint8_t)(val >> 16);
    dst[3] = (uint8_t)(val >> 24);
}

void chacha20_init(chacha20_context *ctx, const uint8_t key[32], const uint8_t nonce[12], uint32_t counter) {
    // Constants
    ctx->state[0] = load32_le((const uint8_t*)constant);
    ctx->state[1] = load32_le((const uint8_t*)constant + 4);
    ctx->state[2] = load32_le((const uint8_t*)constant + 8);
    ctx->state[3] = load32_le((const uint8_t*)constant + 12);

    // Key
    for (int i = 0; i < 8; i++) {
        ctx->state[4 + i] = load32_le(key + i * 4);
    }

    // Counter
    ctx->state[12] = counter;

    // Nonce
    for (int i = 0; i < 3; i++) {
        ctx->state[13 + i] = load32_le(nonce + i * 4);
    }
}

static void chacha20_block(chacha20_context *ctx, uint8_t output[64]) {
    uint32_t x[16];
    memcpy(x, ctx->state, 64);

    for (int i = 0; i < 10; i++) {
        // Column rounds
        QR(x[0], x[4], x[8],  x[12]);
        QR(x[1], x[5], x[9],  x[13]);
        QR(x[2], x[6], x[10], x[14]);
        QR(x[3], x[7], x[11], x[15]);
        
        // Diagonal rounds
        QR(x[0], x[5], x[10], x[15]);
        QR(x[1], x[6], x[11], x[12]);
        QR(x[2], x[7], x[8],  x[13]);
        QR(x[3], x[4], x[9],  x[14]);
    }

    for (int i = 0; i < 16; i++) {
        x[i] += ctx->state[i];
        store32_le(output + i * 4, x[i]);
    }
}

void chacha20_crypt(chacha20_context *ctx, const uint8_t *input, uint8_t *output, size_t len) {
    uint8_t block[64];
    size_t i = 0;

    while (len > 0) {
        chacha20_block(ctx, block);
        
        // Increment counter
        ctx->state[12]++;
        // Note: We don't handle 64-bit counter overflow for state[12] here as per standard 96-bit nonce usage,
        // it just wraps around 32-bit. RFC 7539 uses 32-bit counter.

        size_t n = (len < 64) ? len : 64;
        for (size_t j = 0; j < n; j++) {
            output[i + j] = input[i + j] ^ block[j];
        }

        i += n;
        len -= n;
    }
}

File: js/ChaCha20/ChaCha20.js

// ChaCha20.js
// Ported from c/ChaCha20/chacha20.c

class ChaCha20 {
    constructor(key, nonce, counter = 1) {
        this.state = new Uint32Array(16);
        this.init(key, nonce, counter);
    }

    static ROTL(a, b) {
        return ((a << b) | (a >>> (32 - b))) >>> 0;
    }

    static load32_le(src, offset) {
        return (src[offset] |
               (src[offset + 1] << 8) |
               (src[offset + 2] << 16) |
               (src[offset + 3] << 24)) >>> 0;
    }

    static store32_le(dst, offset, val) {
        dst[offset] = val & 0xFF;
        dst[offset + 1] = (val >>> 8) & 0xFF;
        dst[offset + 2] = (val >>> 16) & 0xFF;
        dst[offset + 3] = (val >>> 24) & 0xFF;
    }

    static hchacha20(key, nonce16) {
        // draft-arciszewski-xchacha-03, Section 2.2
        const st = new Uint32Array(16);
        // Constants "expand 32-byte k"
        st[0] = 0x61707865; st[1] = 0x3320646e; st[2] = 0x79622d32; st[3] = 0x6b206574;
        // Key
        for (let i = 0; i < 8; i++) st[4 + i] = ChaCha20.load32_le(key, i * 4);
        // 128-bit nonce in last 4 words
        for (let i = 0; i < 4; i++) st[12 + i] = ChaCha20.load32_le(nonce16, i * 4);

        const x = new Uint32Array(st);
        for (let i = 0; i < 10; i++) {
            // Column rounds
            ChaCha20.QRstatic(x, 0, 4, 8, 12);
            ChaCha20.QRstatic(x, 1, 5, 9, 13);
            ChaCha20.QRstatic(x, 2, 6, 10, 14);
            ChaCha20.QRstatic(x, 3, 7, 11, 15);
            // Diagonal rounds
            ChaCha20.QRstatic(x, 0, 5, 10, 15);
            ChaCha20.QRstatic(x, 1, 6, 11, 12);
            ChaCha20.QRstatic(x, 2, 7, 8, 13);
            ChaCha20.QRstatic(x, 3, 4, 9, 14);
        }

        // Output: words 0..3 and 12..15, little-endian
        const out = new Uint8Array(32);
        for (let i = 0; i < 4; i++) ChaCha20.store32_le(out, i * 4, x[i]);
        for (let i = 0; i < 4; i++) ChaCha20.store32_le(out, (4 + i) * 4, x[12 + i]);
        return out;
    }

    static QRstatic(x, a, b, c, d) {
        x[a] = (x[a] + x[b]) >>> 0; x[d] ^= x[a]; x[d] = ChaCha20.ROTL(x[d], 16);
        x[c] = (x[c] + x[d]) >>> 0; x[b] ^= x[c]; x[b] = ChaCha20.ROTL(x[b], 12);
        x[a] = (x[a] + x[b]) >>> 0; x[d] ^= x[a]; x[d] = ChaCha20.ROTL(x[d], 8);
        x[c] = (x[c] + x[d]) >>> 0; x[b] ^= x[c]; x[b] = ChaCha20.ROTL(x[b], 7);
    }

    init(key, nonce, counter) {
        // Constants "expand 32-byte k"
        const constant = [0x61707865, 0x3320646e, 0x79622d32, 0x6b206574]; // Little Endian integers
        
        this.state[0] = constant[0];
        this.state[1] = constant[1];
        this.state[2] = constant[2];
        this.state[3] = constant[3];

        // Support:
        // - ChaCha20 (RFC7539): 12-byte nonce
        // - XChaCha20: 24-byte (192-bit) nonce via HChaCha20 subkey derivation
        let realKey = key;
        let realNonce = nonce;
        if (nonce.length === 24) {
            const subkey = ChaCha20.hchacha20(key, nonce.subarray(0, 16));
            const nonce12 = new Uint8Array(12);
            // nonce12 = 0x00000000 || nonce[16..23]
            nonce12.set(nonce.subarray(16, 24), 4);
            realKey = subkey;
            realNonce = nonce12;
        }

        // Key (always 32 bytes)
        for (let i = 0; i < 8; i++) {
            this.state[4 + i] = ChaCha20.load32_le(realKey, i * 4);
        }

        // Counter
        this.state[12] = counter;

        // Nonce
        for (let i = 0; i < 3; i++) {
            this.state[13 + i] = ChaCha20.load32_le(realNonce, i * 4);
        }
    }

    block(output) {
        const x = new Uint32Array(this.state);

        for (let i = 0; i < 10; i++) {
            // Column rounds
            this.QR(x, 0, 4, 8, 12);
            this.QR(x, 1, 5, 9, 13);
            this.QR(x, 2, 6, 10, 14);
            this.QR(x, 3, 7, 11, 15);
            
            // Diagonal rounds
            this.QR(x, 0, 5, 10, 15);
            this.QR(x, 1, 6, 11, 12);
            this.QR(x, 2, 7, 8, 13);
            this.QR(x, 3, 4, 9, 14);
        }

        for (let i = 0; i < 16; i++) {
            x[i] = (x[i] + this.state[i]) >>> 0;
            ChaCha20.store32_le(output, i * 4, x[i]);
        }
    }

    QR(x, a, b, c, d) {
        x[a] = (x[a] + x[b]) >>> 0; x[d] ^= x[a]; x[d] = ChaCha20.ROTL(x[d], 16);
        x[c] = (x[c] + x[d]) >>> 0; x[b] ^= x[c]; x[b] = ChaCha20.ROTL(x[b], 12);
        x[a] = (x[a] + x[b]) >>> 0; x[d] ^= x[a]; x[d] = ChaCha20.ROTL(x[d], 8);
        x[c] = (x[c] + x[d]) >>> 0; x[b] ^= x[c]; x[b] = ChaCha20.ROTL(x[b], 7);
    }

    crypt(input) {
        let inputBytes;
        if (typeof input === 'string') {
            inputBytes = new TextEncoder().encode(input);
        } else {
            inputBytes = input;
        }

        const len = inputBytes.length;
        const output = new Uint8Array(len);
        const blockBuffer = new Uint8Array(64);
        let i = 0;
        let remaining = len;

        while (remaining > 0) {
            this.block(blockBuffer);
            
            // Increment counter
            this.state[12] = (this.state[12] + 1) >>> 0;

            const n = (remaining < 64) ? remaining : 64;
            for (let j = 0; j < n; j++) {
                output[i + j] = inputBytes[i + j] ^ blockBuffer[j];
            }

            i += n;
            remaining -= n;
        }

        return output;
    }
}

module.exports = ChaCha20;
